lib/qx/tool/compiler/utils/Json.js

```diff
@@ -41,15 +43,68 @@ qx.Class.define("qx.tool.compiler.utils.Json", {
       if (!str) {
         return null;
       }
+      // eslint-disable-next-line no-path-concat
+      var qxRoot = __dirname + "/../../../../../";
       var ast = JsonToAst.parseToAst(str);
       var json = JsonToAst.astToObject(ast);
+      var id = this.identifyData(json);
+      if (id) {
+        var ajv = new Ajv({
+          allErrors: true,
+          jsonPointers: true
+        });
+        var schema = "https://raw.githubusercontent.com/qooxdoo/qooxdoo-compiler/master/resource/" + id.type + "-" + String(id.version) + ".json";
+        var valid = ajv
+          .addSchema(require(qxRoot + "resource/compile-" + String(id.version) + ".json"))
+          .addSchema(require(qxRoot + "resource/Manifest-" + String(id.version) + ".json"))
+          .validate(schema, json);
+        if (!valid) {
+          const message = betterAjvErrors(schema, json, ajv.errors, {format: "cli", indent: 2});
+          console.warn("WARNING: "+id.type + ".json\n" + message);
+          if (id.version > 0) {
+            const err = betterAjvErrors(schema, json, ajv.errors, {format: "js"});
+            throw new Error(err[0].error);
+          }
+        }
+      }
       return json;
     },
-    
+    /**
+     * Identify Type and Schema Version
+     *
+     * @param data {Object} Manifest.json or compile.json
+     * @return {Object?} object with data and version key
+     */
+
+    identifyData: function(data) {
+      if (data.$schema) {
+        let match = data.$schema.match(/([^-/\s]+)-(\d+)\.json$/);
+        if (match) {
+          return {
+            type: match[1],
+            version: parseInt(match[2])
+          };
+        }
+      }
+      if (data.applications) {
```

Comment:
I'm a bit concerned about using information in the file to identify the file type. I'd feel better if the file type were passed in. Consider, for example, an otherwise completely valid compile.json file where the user just forgot to add the `applications` section. It wouldn't handle it as a Manifest file unless it had `provides` in it (but handling it as a Manifest file would show all sorts of errors since it isn't a Manifest file). It would return null, i.e., no known type, and you wouldn't be told the one thing that was wrong with your compile.json file.

Reply:
yes ... I was about to revamp the whole api there ... and require specification of the filetype in the reader call